### PR TITLE
Track the execution payload of optimisically sync'd, finalized merge transition block

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
So that we can validate the terminal block conditions when the EL is back online, keep a record of the execution payload for the merge transition block if it was optimistically verified and has been finalized. On disk we only store the slot of the merge block and can reload the ExecutionPayload from that.

Not done yet is adding the ability to clear the recorded payload when it has later been verified (or to actually verify it later...)

## Fixed Issue(s)
#4671 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: storage/src/main/java/tech/pegasys/teku/storage/store/Store.java [hunk 1]
```diff
@@ -87,12 +89,12 @@
   Checkpoint justifiedCheckpoint;
   Checkpoint bestJustifiedCheckpoint;
   UInt64 latestValidFinalizedSlot = UInt64.ZERO;
+  Optional<ExecutionPayload> finalizedOptimisticTransitionPayload = Optional.empty();
```
Comment (tbenr) [inline storage/src/main/java/tech/pegasys/teku/storage/store/Store.java]: nit: `finalizedOptimisticTransitionPayload` is always initialized in the constructor, so initialization here is redundant.

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/BlockMetadataStore.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ void applyUpdate(
   interface HaltableNodeProcessor {
 
     static HaltableNodeProcessor fromNodeProcessor(final NodeProcessor processor) {
-      return (child, slot, parent) -> {
+      return (child, slot, parent, executionHash) -> {
```
Comment (tbenr) [inline protoarray/src/main/java/tech/pegasys/teku/protoarray/BlockMetadataStore.java:46]: nit: are we using `__` on non-used lambda params?

File: storage/src/main/java/tech/pegasys/teku/storage/store/Store.java [hunk 3]
```diff
@@ -625,7 +641,7 @@ VoteTracker getVote(UInt64 validatorIndex) {
     try {
       forkChoiceStrategy.processHashesInChainWhile(
           blockRoot,
-          (root, slot, parent) -> {
+          (root, slot, parent, executionHash) -> {
```
Comment (tbenr) [inline storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:644]: nit: same here

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java [hunk 4]
```diff
@@ -607,6 +623,25 @@ private void updateFinalizedData(
...
               .collect(Collectors.toSet()));
     }
+
+    return optimisticTransitionPayload;
+  }
+
+  private Optional<ExecutionPayload> updateFinalizedOptimisticTransitionBlock(
+      final boolean isFinalizedOptimisticBlockRootSet,
+      final Optional<Bytes32> finalizedOptimisticTransitionBlockRoot) {
+    if (isFinalizedOptimisticBlockRootSet) {
```
Comment (tbenr) [inline storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java:633]: This is the reason why we need an additional boolean aside the Optional, to actually do put\delete only once.
So on DB level we can potentially remove the optimistic transition block slot. Which I guess will be done when the actual validation will be done.

**Review by tbenr — APPROVED** [review]
LGTM. some nits and personal reflections :)

File: protoarray/src/main/java/tech/pegasys/teku/protoarray/BlockMetadataStore.java (hunk 2 above)
Comment (ajsutton) [reply protoarray/src/main/java/tech/pegasys/teku/protoarray/BlockMetadataStore.java:46]: I'd only use that when it's the only option, and particularly then when it's actually a `Void`.  I think it makes sense to use the appropriate name for things that do have meaning.

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java (hunk 4 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java:633]: Yep exactly - the value itself may be present or absent so we either need a separate boolean to track whether it's set or to use `Optional<Optional<Bytes32>>` and the separate boolean seemed a lot neater and easier to understand.

File: storage/src/main/java/tech/pegasys/teku/storage/store/Store.java (hunk 1 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/store/Store.java]: Good catch.

**Pull request merged**